### ISSUE TITLE
Using system file dialogs for mounting removable media

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#2719]
One common complaint that comes up from time to time is the difficulty of mounting removable media via the config. The preferable way would be to just select the file using an OS file dialog, like normal people tend to do 😛 This would be an especially welcome addition when you just want to quickly test something, without making the mount permanent in your config.

While the OSD will happen unless I'm hit by an evil meteor, there's no way I'm going to write and maintain a full-fledged, cross-platform file selector dialog in our future custom UI. If selecting files would be a very common operation, I might consider it (begrudgingly). But thankfully, that's not the case, so we can simply do whst other renowned emulators such as WinUAE, VICE and FUSE are doing--just pop up the OS file selection dialog and call it a day.

There's a handy little (tiny!) library that does the job perfectly. I've been using it with great succes in my own project, Gridmonger:

https://github.com/AndrewBelt/osdialog

The idea is simple:

- Create new shortcut for assigning media to the first CD-ROM drive, which would open an OS file selector dialog.
- Maybe a second shortcut for the second CD drive, but support for all drives should probably only be added via the OSD where you can just click on the available drive buttons.
- Similar shortcuts for the first and second floppy drives.

Well, that's it! 

Saving the current state of the mounts into the local config is out of scope for now; this is only for quick testing. Part of the OSD project will be persisting config changes made via the GUI into the configs, but that's a much bigger task.

This should be nice low hanging fruit to pick for a quick win in the interim and the changes won't need to be reverted when the full OSD enters the picture (oh, the puns! 🤣 )

**Comment by weirddan455** [comment]
This is going to be a bit of a pain point on Linux since Linux doesn't have one standard file picker (unlike on Windows where you could call WIN32 API and be done with it).

That osdialog library you mentioned requires linking with GTK which is a pretty heavy dependency and isn't the "native" file picker for all environments, namely KDE uses its own Qt based file picker.

That's not necessarily a deal breaker.  Firefox and Chromium for example decided to not care about that and just use GTK.  We may also want to consider making this an optional compile time feature so we don't make GTK a required dependency, although most Linux users probably have GTK installed so that's debatable.

I am curious what your plans are for future GUI features.  If you wanted to add a full featured GUI, I would definitely recommend Qt over GTK.  Qt has a much nicer C++ API and works really well on Windows and Mac.  I've only heard bad things about GTK for cross platform and it uses an ugly C API with its Glib dependency that tries to ram object oriented features down C's throat.

Mini GTK rant aside(I don't like that library)... these are some things worth considering.  If we're _only_ using GTK for a file picker, it's probably fine though I worry a little bit about adding extra dependencies and making our builds more unwieldy.

**Comment by johnnovak** [comment]
Gridmonger example — Ctrl-O just brings up the system open file dialog

<img width="1728" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/dd049e42-01fe-483a-813e-f7af1deac894">

**Comment by kcgen** [comment]
Single-file, cross-platform, practical solution - very nice find!

**Comment by johnnovak** [comment]
> This is going to be a bit of a pain point on Linux since Linux doesn't have one standard file picker (unlike on Windows where you could call WIN32 API and be done with it).
> 
> That osdialog library you mentioned requires linking with GTK which is a pretty heavy dependency and isn't the "native" file picker for all environments, namely KDE uses its own Qt based file picker.

The osdialog lib was written by the guy who single-handedly created [VCVRack](https://vcvrack.com/), which is as professional as it gets for an open-source app that runs on Linux (in addition to Windows and macOS). So I kinda thought the GTK dependency is okay.

We have debated the merits of dynamic linking recently (I was more like trolling... okay :trollface:), so I'm guessing most people will have GTK installed as a dynamic library by default anyway, and if not... I guess bad luck? 😄 Seriously, I'm not a desktop Linux user, I'm just going by the VCVRack example and the guy seems to know very much what he's doing in all aspects of the project. I'm personally happy to go with the GTK dependency as a dynamic lib if 98% of people have that on their Linux machines anyway, then the 2% can install it begrudgingly if they want the dialog, otherwise go use the config file approach 🤷🏻 

> I am curious what your plans are for future GUI features. If you wanted to add a full featured GUI, I would definitely recommend Qt over GTK. Qt has a much nicer C++ API and works really well on Windows and Mac. I've only heard bad things about GTK for cross platform and it uses an ugly C API with its Glib dependency that tries to ram object oriented features down C's throat.
> 
> Mini GTK rant aside(I don't like that library)... these are some things worth considering. If we're _only_ using GTK for a file picker, it's probably fine though I worry a little bit about adding extra dependencies and making our builds more unwieldy.

Thanks, but I'm not going to touch Qt with a ten-feet pole for this particular job 😅 (or any generic toolkit; I'm an equal-opportunity toolkit-disliker 😎🤘🏻). 

The TL;DR is that the OSD will be overlaid on top of the rendered output, like the UI of a TV or [Kodi](https://kodi.tv/). It will be OpenGL drawn, completely custom, and it will look nice 😄 The immediate mode "mini toolkit" will be completely custom, written by me—it's gonna be a kind of a port of my 10k LOC GUI library I specifically wrote for Gridmonger.

You can read more about my ideas for the OSD here:
https://github.com/dosbox-staging/dosbox-staging/discussions/1953

And about my general "philosophy" about cross-platform UIs and GUIs in general in the blog post I wrote in 2016. I can assure you the situation hasn't changed _one bit_ since then:
https://blog.johnnovak.net/2016/05/29/cross-platform-gui-toolkit-trainwreck-2016-edition/

**Comment by weirddan455** [comment]
>  Thanks, but I'm not going to touch Qt with a ten-feet pole for this particular job sweat_smile (or any generic toolkit; I'm an equal-opportunity toolkit-disliker sunglasses🤘🏻).

Yeah, I fully understand that.  Hence my slight concern about linking with GTK.  It is a pretty standard thing to do on Linux.  Even if I don't like it, I'm not sure there is a better a option offhand, though I could do a bit of research later.  Like I said, not necessarily a dealbreaker, just some potential downsides to be aware of.

>  The TL;DR is that the OSD will be overlaid on top of the rendered output, like the UI of a TV or [Kodi](https://kodi.tv/). It will be OpenGL drawn, completely custom, and it will look nice smile The immediate mode "mini toolkit" will be completely custom, written by me—it's gonna be a kind of a port of my 10k LOC GUI library I specifically wrote for Gridmonger.

That sounds great man.  Can't wait to test that out.

>  And about my general "philosophy" about cross-platform UIs and GUIs in general in the blog post I wrote in 2016. I can assure you the situation hasn't changed one bit since then:
https://blog.johnnovak.net/2016/05/29/cross-platform-gui-toolkit-trainwreck-2016-edition/

Looks like you've been down this road before.  I'll give that full read later.

>  PS: Statically linking GTK in this particular case on Linux would be abysmal, I agree with that—I'd never do that. But I think with dynamic linking it's okay? The alternative is that I will implement a cross-platform file browser weary But this ticket was about quick wins.

I don't know if you need to go that far.  It would fit in better with the OSD you're describing, rather than have a system pop-up but it could also be a lot of work.  On Windows (and presumably Mac), it's fairly easy to just make a system call to show a file picker so if you're fine with that approach for those systems, I'm sure we can make do with something similar on Linux :smile:

One thought I had was to do a runtime `dlopen` to avoid having to link at compile time (and if the user doesn't have GTK, we just log an error), though that might have its own issues.

**Comment by johnnovak** [comment]
> One thought I had was to do a runtime `dlopen` to avoid having to link at compile time (and if the user doesn't have GTK, we just log an error), though that might have its own issues.

Yeah, something like that could work, perhaps. Good idea. I guess I don't like the option to _not_ support this on Linux; features should be available on all supported platforms and we should only deviate from the exceptional circumstances. Select a file via a dialog is a pretty standard thing, far from exceptional 😄

**Comment by rderooy** [comment]
While I don't have any issue with the GTK part of this, would using such a toolkit not force you out of fullscreen mode? It also would prevent it from working when you don't have a Wayland or X11 display. e.g. running something like RetroPie on a Raspberry Pi?

FYI, dosbox-x uses https://sourceforge.net/projects/tinyfiledialogs/

**Comment by johnnovak** [comment]
> While I don't have any issue with the GTK part of this, would using such a toolkit not force you out of fullscreen mode? It also would prevent it from working when you don't have a Wayland or X11 display. e.g. running something like RetroPie on a Raspberry Pi?

Yeah good point, I guess I'd need to experiment with it. On Windows at least borderless fullscreen is the standard for a while now, so the dialog would just appear on top of your "fullscreen" image. It's actually quite a challenge to get Windows into "true" fullscreen" mode these days, and mostly pointless, actually, given all the drawbacks.

RetroPie and similar — I guess those are kinda "edge" use cases anyway, and the config based mounting won't go away. So on one hand maybe not such a big loss, and you use those devices mostly with game controllers from the couch anyway, right? So you kinda rely on pre-packed gamepacks, that's what I assume.

But yeah, this kinda all nudges me towards just writing my own within the OSD framework, after all...

> FYI, dosbox-x uses https://sourceforge.net/projects/tinyfiledialogs/

Yeah, had a quick look and I have a bad feeling about this... Not so "tiny", tries to do too much, I actually find it quick bloated. I'd actually much rather suck it up and write my own than using this. `osdialog` is quite minimal, that's why I like it.

**Comment by johnnovak** [comment]
I guess we can all just try https://vcvrack.com/ on different platforms. Free download, have at it! 😎

**Comment by johnnovak** [comment]
Ok... on macOS if I go to fullscreen with F11 then hit Cmd+O to bring up the open file dialog, then first it minimizes itself to the tray, pops up the dialog... then on close does the reverse. LOL 😄 

Not terrible, certainly usable, but I wouldn't call it fantastic either 😅 

This is a useful discussion, though, it might convince me that the OSD-based file browser is the way to go.

**Comment by rderooy** [comment]
Wayland also does not allow you to go into true fullscreen mode. There is no way for a program to set its own resolution or refresh rate, and this is on purpose.

This also applies to programs using XWayland, they will only see a single resolution and refresh rate, which is the one for the monitor they are on. Attempts to go into fullscreen will just result just give them a full-screen border-less window.
